jni_bindings/src/lib.rs

```diff
@@ -73,159 +69,226 @@ pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_00024Comp
-    // only u32 should be passed as node
-    let node = u32::try_from(node).expect("value beyond `u32` range");
-    let node = NodeRef::new(node as usize);
-    let dom = unsafe { jlong_to_pointer::<dom::Document>(this).as_ref().unwrap() };
-    let mut buf = String::new();
-
-    if let Err(err) = dom.print_node(node, &mut buf, dom::PrintOptions::Pretty) {
-        let exception_class = env.find_class("java/lang/RuntimeException").unwrap();
-        let message = format!("Rust panic occurred in node_to_string: {:?}", err);
-        env.throw_new(exception_class, message).unwrap();
+    let this = this as *mut Document;
+
+    if this.is_null() {
+        env.throw_new(
+            "java/lang/NullPointerException",
+            "Document::node_to_string called with `this` as null pointer",
+        )
+        .unwrap();
+
         JObject::null().into()
     } else {
-        from_std_string_jstring(buf, env)
+        let node_ref = NodeRef::new(node_ref as usize);
+        let doc = &*this;
+        let mut buf = String::new();
+
+        if let Err(err) = doc.print_node(node_ref, &mut buf, dom::PrintOptions::Pretty) {
+            let message = format!("Document::node_to_string failed: {:?}", err);
+            env.throw_new("java/lang/RuntimeException", message)
+                .unwrap();
+
+            JObject::null().into()
+        } else {
+            env.new_string(buf)
+                .unwrap_or_else(|_| JObject::null().into())
+        }
     }
 }
 
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_root<'local>(
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_root<'local>(
     mut env: JNIEnv<'local>,
     _: JClass<'local>,
+    // non-null raw pointer to Document
     this: jlong,
+    // NodeRef of root
 ) -> jint {
-    let result = catch_unwind(|| unsafe {
-        let dom: &dom::Document = jlong_to_pointer::<dom::Document>(this).as_ref().unwrap();
-        dom.root().as_u32() as jint
-    });
-
-    match result {
-        Ok(jint) => jint,
-        Err(err) => {
-            let exception_class = env.find_class("java/lang/RuntimeException").unwrap();
-            let message = format!("Rust panic occurred in Document_root: {:?}", err);
-            env.throw_new(exception_class, message).unwrap();
-            0
-        }
+    let this = this as *mut Document;
+
+    if this.is_null() {
+        env.throw_new(
+            "java/lang/NullPointerException",
+            "Document::root called with `this` as null pointer",
+        )
+        .unwrap();
+
+        0
+    } else {
+        let doc = &*this;
+        doc.root().as_u32() as jint
     }
 }
 
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node<'local>(
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node<'local>(
     mut env: JNIEnv<'local>,
     _: JClass<'local>,
+    // non-null raw pointer to Document
```

Comment:
The `non-null` can't be true or there would be no need for the `this.is_null()` check below.
```suggestion
    // raw pointer to Document
```